refactor(config): enable incremental TypeScript builds

zharinov opened the pull request:
## Changes

Enable incremental TypeScript compilation for faster `pnpm type-check`:

| Scenario | Before | After |
|----------|--------|-------|
| Warm cache | 1:44 | ~9s |
| After file change | 1:44 | ~12s |

- Add `incremental: true` to tsconfig
- Store build info in `.cache/tsbuildinfo` (already gitignored)
- Cache tsbuildinfo in CI for faster runs

## Context

- [x] This doesn't close an Issue, but I accept the risk that this PR may be closed if maintainers disagree with its opening or implementation

## AI assistance disclosure

- [x] Yes — substantive assistance (AI-generated non‑trivial portions of code, tests, or documentation).

## Documentation (please check one with an [x])

- [x] No documentation update is required

## How I've tested my work (please select one)

- [x] Code inspection only

zharinov committed:
refactor(config): enable incremental TypeScript builds

Enable TypeScript incremental compilation to dramatically improve
type-check performance for local development.

Performance improvements:
- Cold cache: 1:44 → 1:05 (~40% faster)
- Warm cache: 1:44 → ~9s (~10x faster)
- After file change: 1:44 → ~12s (~8x faster)

The build info is stored in `.cache/tsbuildinfo` which is already
gitignored. IDE/LSP is unaffected as tsserver maintains its own
in-memory project graph.

zharinov committed:
chore(ci): cache tsbuildinfo for faster type-check

Add cache restore/save for the TypeScript incremental build info
to speed up type-check in CI. Cache is keyed by pnpm-lock.yaml hash
and saved on push events.

jamietanna reviewed (dismissed):
Looks good! @renovatebot/maintainers either of you want to take a look before we enable this?

secustor commented on .github/workflows/build.yml:
```diff
@@ -326,9 +326,23 @@ jobs:
+        uses: actions/cache/restore@0057852bfaa89a56745cba8c7296529d2fc39830 # v4.3.0
+        with:
+          path: .cache/tsbuildinfo
+          key: tsbuildinfo-${{ hashFiles('pnpm-lock.yaml') }}
```
```suggestion
          key: tsbuildinfo-${{ hashFiles('pnpm-lock.yaml') }}-${{ github.sha }}
```
Shouldn't this match the save below

zharinov committed:
Update .github/workflows/build.yml

Co-authored-by: Sebastian Poxhofer <secustor@users.noreply.github.com>

zharinov committed:
fix(ci): align tsbuildinfo cache keys with eslint/prettier pattern

Use `run_id`/`run_attempt` for save key uniqueness instead of `github.sha`
to match the existing cache key patterns used for eslint and prettier.

zharinov commented:
Fixed cache keys, now key usage is consistent with other CI jobs.
It uses key that never matches (it's a trick), and restore-key gets the value from `main` or from the previous run.
